server/channels/app/post.go

```diff
@@ -741,85 +740,89 @@ func (a *App) UpdatePost(c *request.Context, post *model.Post, safeUpdate bool)
-	}
-	message.Add("post", postJSON)
 
-	published, err := a.publishWebsocketEventForPermalinkPost(c, rpost, message)
+	err = a.publishWebsocketEventForPost(c, rpost, message)
 	if err != nil {
 		return nil, err
 	}
-	if !published {
-		a.Publish(message)
-	}
 
 	a.invalidateCacheForChannelPosts(rpost.ChannelId)
 
 	return rpost, nil
 }
 
-func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.Post, message *model.WebSocketEvent) (published bool, err *model.AppError) {
-	var previewedPostID string
-	if val, ok := post.GetProp(model.PostPropsPreviewedPost).(string); ok {
-		previewedPostID = val
-	} else {
-		return false, nil
+// publishWebsocketEventForPost publishes the websocket event only for post create/edit.
+// The cases of post delete/unread does not need special handling as they don't bother
+// with the post content.
+//
+// This method assumes that if there's a permalink, it's already attached to the post.
+// If the user doesn't have access then this method will wipe that off.
+func (a *App) publishWebsocketEventForPost(c request.CTX, post *model.Post, message *model.WebSocketEvent) *model.AppError {
+	postJSON, jsonErr := post.ToJSON()
+	if jsonErr != nil {
+		return model.NewAppError("publishWebsocketEventForPost", "app.post.marshal.app_error", nil, "", http.StatusInternalServerError).Wrap(jsonErr)
 	}
+	message.Add("post", postJSON)
+
+	defer a.Publish(message)
 
-	if !model.IsValidId(previewedPostID) {
-		mlog.Warn("invalid post prop value", mlog.String("prop_key", model.PostPropsPreviewedPost), mlog.String("prop_value", previewedPostID))
-		return false, nil
+	permalinkPreviewedPost := post.GetPreviewPost()
+	if permalinkPreviewedPost == nil {
+		return nil
 	}
 
-	previewedPost, err := a.GetSinglePost(previewedPostID, false)
-	if err != nil {
-		if err.StatusCode == http.StatusNotFound {
-			mlog.Warn("permalinked post not found", mlog.String("referenced_post_id", previewedPostID))
-			return false, nil
-		}
-		return false, err
+	if !model.IsValidId(permalinkPreviewedPost.PostID) {
+		mlog.Warn("invalid preview post ID", mlog.String("prop_value", permalinkPreviewedPost.PostID))
+		return nil
 	}
 
-	channelMembers, err := a.GetChannelMembersPage(c, post.ChannelId, 0, 10000000)
+	// To remain secure by default, we wipe out the metadata unconditionally.
+	post.Metadata.Embeds[0].Data = nil
+	postWithoutPermalinkPreviewJSON, err := post.ToJSON()
 	if err != nil {
-		return false, err
+		return model.NewAppError("publishWebsocketEventForPost", "app.post.marshal.app_error", nil, "", http.StatusInternalServerError).Wrap(jsonErr)
 	}
 
-	permalinkPreviewedChannel, err := a.GetChannel(c, previewedPost.ChannelId)
-	if err != nil {
-		if err.StatusCode == http.StatusNotFound {
-			mlog.Warn("channel containing permalinked post not found", mlog.String("referenced_channel_id", previewedPost.ChannelId))
-			return false, nil
+	previewedPost, appErr := a.GetSinglePost(permalinkPreviewedPost.PostID, false)
```

Comment:
There is a race here. If the permalinked post gets deleted after entering `publishWebsocketEventForPost` but before reaching this line, `GetSinglePost` here will return an error, `publishWebsocketEventForPost`, and the deferred `Publish` runs without setting the `ChannelHook`. Which means the deleted post's contents get broadcast to all users in the channel unsanitized.

Reply:
> If the permalinked post gets deleted after entering publishWebsocketEventForPost but before reaching this line

The `createPost` endpoint doesn't return until it returns from `publishWebsocketEventForPost` and finishes the other stuff. Were you thinking of a case where the DB is manually modified?

Would it be safer for you if we move the `message.Add("post", postJSON)` to before L804? Just to prevent sending a post with full metadata?

Reply:
Why does it matter when the `createPost` endpoint returns? This is the case I'm thinking about:

1. User `A` creates post `a` in a private channel
2. User `B` creates post `b` in a public channel and links to post `a`
3. Simultaneously with step `2`, user `A` deletes post `a`
4. All users in the public channel receive a WS broadcast about the creation of post `b`, containing the embedded post `b`

`createPost` returns some time after step `4` but at that point the WS broadcast has already gone out.

Reply:
Ah yes, sorry I had a brain fart. So we have two options: 
- Either don't publish in case there's an error
- Move the `message.Add` line below these checks.

I am leaning towards option 1 as that looks more in line with the earlier implementation.

Reply:
A third option would be to set the `ChannelHook` early. The variables it uses, like `permalinkPreviewedChannel`, would have to be initialized with empty values, but having them empty when the function returns early is a safe failure mode.

I'm ok with any of the three options.